auth/auth.go

```diff
@@ -157,31 +157,56 @@ func (auth *Authenticator) getPrincipal(docID string, factory func() Principal)
 func (auth *Authenticator) rebuildChannels(princ Principal) error {
 	channels := princ.ExplicitChannels().Copy()
 
-	// Changes for vbucket sequence management.  We can't determine relative ordering of sequences
-	// across vbuckets. To avoid redundant channel backfills during changes processing, we maintain
-	// the previous vb/seq for a channel in PreviousChannels.  If that channel is still present during
-	// this rebuild, we reuse the vb/seq from PreviousChannels (using UpdateIfPresent).  If PreviousChannels
-	// is nil, reverts to normal sequence handling.
-
-	previousChannels := princ.PreviousChannels().Copy()
-	if previousChannels != nil {
-		channels.UpdateIfPresent(previousChannels)
-	}
-
 	if auth.channelComputer != nil {
 		viewChannels, err := auth.channelComputer.ComputeChannelsForPrincipal(princ)
 		if err != nil {
 			base.Warnf("channelComputer.ComputeChannelsForPrincipal returned error for %v: %v", base.UD(princ), err)
 			return err
 		}
-		if previousChannels != nil {
-			viewChannels.UpdateIfPresent(previousChannels)
-		}
 		channels.Add(viewChannels)
 	}
 	// always grant access to the public document channel
 	channels.AddChannel(ch.DocumentStarChannel, 1)
 
+	var previousChannelEntries ch.TimedSet
+	var previousInvalSeq uint64
+
+	if princ.PreviousChannels() != nil {
+		previousChannelEntries = princ.PreviousChannels().Entries
+		previousInvalSeq = princ.PreviousChannels().InvalSeq
+	}
+
+	removedChannels := ch.TimedSet{}
+	for previousChannelName, previousChannelInfo := range previousChannelEntries {
+		if _, ok := channels[previousChannelName]; !ok {
+			removedChannels[previousChannelName] = previousChannelInfo
+		}
+	}
+
+	channelHistory := princ.ChannelHistory()
```

Comment:
What's the reason to not just stick previousChannels into channelHistory at invalidation time, instead of waiting for next auth?

Reply:
Couple reasons:
- Want to know what current assignments are so we can diff with them so find changes. The query / work done to calculate current assignments is already done in rebuild.
- In the case of multiple invalidations you don't want to do all the work until necessary.

To have work done on invalidation:
- Need a way to find current assignments, involving adding a query
- Lose the advantage of only doing the invalidation once. You'd need to recalculate every time there is a change.

Reply:
Addressed?